File: js/day-1.js
const fs = require("fs")

const input = fs
    .readFileSync("inputs/day-1.txt", "utf8")
    .split("\n")
    .map(_ => parseInt(_, 10))

const isGreaterThanPrevious = (acc, cur, i, arr) => {
    if (i === 0) return acc
    else if (cur > arr[i - 1]) return acc + 1
    else return acc
}

const getRollingAverage = (_, i, arr) => {
    if (!arr[i + 2]) return null
    else return arr[i] + arr[i + 1] + arr[i + 2]
}

const part1 = input
    .reduce(isGreaterThanPrevious, 0)

const part2 = input
    .flatMap(getRollingAverage)
    .reduce(isGreaterThanPrevious, 0)

console.log("part 1:", part1)
console.log("part 2:", part2)

File: go/day-1.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"strings"
)

func check(e error) {
	if e != nil {
		panic(e)
	}
}

func strToIntSlice(s string) ([]int, error) {
	var result []int
	rawJson := "[" + strings.ReplaceAll(s, "\n", ",") + "]"
	err := json.Unmarshal([]byte(rawJson), &result)

	return result, err
}

func getInput() ([]int, error) {
	rawInput, err := os.ReadFile("inputs/day-1.txt")
	check(err)

	return strToIntSlice(string(rawInput))
}

func getRollingAverage(input []int) ([]int) {
	var result []int

	for i := 0; i < len(input) - 2; i++ {
		result = append(result, input[i] + input[i + 1] + input[i + 2])
	}

	return result
}

func part1() int {
	input, err := getInput()
	check(err)

	count := 0

	for i := 1; i < len(input); i++ {
		if input[i] > input[i - 1] {
			count++
		}
	}

	return count
}

func part2() int {
	input, err := getInput()
	check(err)
	rollingAverage := getRollingAverage(input)

	count := 0

	for i := 1; i < len(rollingAverage); i++ {
		if rollingAverage[i] > rollingAverage[i - 1] {
			count++
		}
	}

	return count
}

func main() {
	fmt.Println("part 1: ", part1())
	fmt.Println("part 2: ", part2())
}
